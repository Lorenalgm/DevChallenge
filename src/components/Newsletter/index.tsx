import {
  Flex,
  FlexProps,
  Heading,
  Text,
  Box,
  useToast,
  useBoolean
} from '@chakra-ui/react'
import { Player } from '@lottiefiles/react-lottie-player'

import NewsletterForm from 'components/NewsletterForm'

import NewsletterResources from 'services/resources/newsletter'

const Newsletter = ({ ...props }: FlexProps) => {
  const toast = useToast()
  const [isLoading, setIsLoading] = useBoolean()

  const handleSubmit = async (email: string) => {
    setIsLoading.on()
    const { error } = await NewsletterResources.saveEmail(email)
    setIsLoading.off()

    if (error) {
      toast({
        title: 'Opa, algo deu errado! Pode tentar novamente? :c',
        status: 'error',
        position: 'top-right',
        isClosable: true
      })

      return
    }

    toast({
      title: 'Feito! Você será o primeiro a saber sobre novos desafios :)',
      status: 'success',
      position: 'top-right',
      isClosable: true
    })
  }

  return (
    <Flex
      marginX="auto"
      alignItems="center"
      flexDirection={['column', 'column', 'row']}
      {...props}
    >
      <Box
        as={Player}
        autoplay={true}
        controls={false}
        loop={true}
        src="https://assets5.lottiefiles.com/private_files/lf30_WdTEui.json"
        height="300px"
        width="300px"
        aria-label="developer video"
      />

      <Flex flexDirection="column" ml={[0, 0, '6', '10']}>
        <Heading as="h3" fontSize={['2xl', '3xl', '3xl', '4xl']}>
          Seja notificado sobre novos desafios!
        </Heading>

        <Text color="primary.500" fontSize={['md', 'lg', 'lg', 'xl']}>
          Inscreva-se para ser o primeiro a saber sobre novos desafios :)
        </Text>

        <NewsletterForm mt="6" isLoading={isLoading} onSubmit={handleSubmit} />
      </Flex>
    </Flex>
  )
}

export default Newsletter
