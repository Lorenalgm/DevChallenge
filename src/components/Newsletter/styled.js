import styled from 'styled-components';
import media from 'styled-media-query';

export const NewsletterContainer = styled.section`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;

  ${media.lessThan('medium')`
        flex-direction: column;
    `}

  div {
    display: flex;
    flex-direction: column;
    align-items: left;
    margin-left: 2em;

    ${media.lessThan('medium')`
            align-items: center;
            justify-content: center;
            text-align: center;
            margin-left: 0;
        `}
  }
`;

export const NewsletterTitle = styled.h1`
  color: var(--white);
  font-size: 2em;
  margin-bottom: 0.1em;

  ${media.lessThan('medium')`
        font-size: 1.3em;
        margin-bottom: 0.1em;
    `}
`;

export const NewsletterParagraph = styled.p`
  color: var(--yellow);
  font-size: 1em;

  ${media.lessThan('medium')`
        margin: 1em;
    `}
`;

export const NewsletterForm = styled.form`
  display: flex;
  flex-direction: column;

  ${media.lessThan('medium')`
        display: flex;
        align-content: center;
        justify-content: center;
        text-align: center;
        margin-bottom: 3em;
    `}

  input {
    height: 3.5em;
    width: 20em;
    margin-top: 1em;
    margin-bottom: 1em;
    border-radius: 0.5em;
    padding: 20px;
    font-size: 18px;

    ${media.lessThan('medium')`
            height: 3em;
            width: 18em;
        `}
  }

  button {
    color: var(--white);
    background-color: var(--purple);
    width: 350px;
    height: 60px;
    font-weight: bold;
    font-size: 20px;
    border-radius: 50px;
    cursor: pointer;
    transition: 0.3s;

    ${media.lessThan('medium')`
            width: 16em;
            height: 3em;
        `}

    &:hover {
      background-color: var(--dark-purple);
    }
  }
`;
