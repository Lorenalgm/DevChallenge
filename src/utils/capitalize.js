export default function capitalize(word) {
    if (!word) return undefined;

    const firstLetter = word[0];
    const restOfWord = word.slice(1);

    return firstLetter.toUpperCase() + restOfWord;
}
