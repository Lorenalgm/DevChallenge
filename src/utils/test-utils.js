import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';
import { render } from '@testing-library/react';

import { GlobalStyles } from '../styles/GlobalStyles';

const customRender = (component, { ...renderOptions } = {}) =>
    render(
        <>
            <GlobalStyles />
            {component}
            <ToastContainer />
        </>,
        { wrapper: MemoryRouter, ...renderOptions }
    );

export * from '@testing-library/react';
export { customRender as render };
