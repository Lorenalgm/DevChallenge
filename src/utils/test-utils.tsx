import { render, RenderResult } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'
import { matchers } from '@emotion/jest'

expect.extend(matchers)

import theme from 'theme'

export const customRender = (children: React.ReactNode): RenderResult =>
  render(
    <ChakraProvider resetCSS theme={theme}>
      {children}
    </ChakraProvider>
  )

export * from '@testing-library/react'
export { customRender as render }
