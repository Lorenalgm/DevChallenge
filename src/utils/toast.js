import { toast } from 'react-toastify';

const messages = {
  error(msg) {
    return msg || 'Ocorreu um erro no processamento';
  },
  success(msg) {
    return msg || 'Sucesso';
  },
};

const ToastNotification = {
  notify: (method, msg) => {
    const methodStr = method;
    method = toast[method];
    method(messages[methodStr](msg));
  },
};

export default ToastNotification;
