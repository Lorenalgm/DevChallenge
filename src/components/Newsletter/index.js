import React, { useState } from 'react';
import { Player } from '@lottiefiles/react-lottie-player';
import api from '../../services/api';
import ToastNotification from '../../utils/toast';

import * as S from './styled';

export default function Newsletter() {
  const [email, setEmail] = useState('');

  async function handleSubscribe(e) {
    e.preventDefault();

    const data = {
      email,
    };

    try {
      await api.post('newsletter', data);
      ToastNotification.notify(
        'success',
        'Feito! Você será o primeiro a saber sobre novos desafios :)'
      );
    } catch (err) {
      ToastNotification.notify(
        'error',
        'Opa, algo deu errado! Pode tentar novamente? :c'
      );
    }
  }

  return (
    <S.NewsletterContainer>
      <div>
        <Player
          autoplay={true}
          controls={false}
          loop={true}
          src="https://assets5.lottiefiles.com/private_files/lf30_WdTEui.json"
          style={{ height: '300px', width: '300px' }}
        />
      </div>
      <div>
        <S.NewsletterTitle>
          Seja notificado sobre novos desafios!
        </S.NewsletterTitle>
        <S.NewsletterParagraph>
          Inscreva-se para ser o primeiro a saber sobre novos desafios :)
        </S.NewsletterParagraph>
        <S.NewsletterForm onSubmit={handleSubscribe}>
          <input
            type="email"
            placeholder="E-mail"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            required
          />
          <button type="submit">Inscrever</button>
        </S.NewsletterForm>
      </div>
    </S.NewsletterContainer>
  );
}
