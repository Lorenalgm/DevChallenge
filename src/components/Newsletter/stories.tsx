import { Story, Meta } from '@storybook/react/types-6-0'

import Newsletter from '.'

export default {
  title: 'Newsletter',
  component: Newsletter
} as Meta

export const Default: Story = () => <Newsletter />
