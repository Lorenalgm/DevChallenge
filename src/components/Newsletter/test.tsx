import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import Newsletter from '.'

import NewsletterResources from 'services/resources/newsletter'
const saveEmail = jest.spyOn(NewsletterResources, 'saveEmail')

describe('<Newsletter />', () => {
  it('should render correctly', () => {
    const { container } = render(<Newsletter />)

    expect(
      screen.getByRole('heading', {
        name: /seja notificado sobre novos desafios!/i
      })
    ).toBeInTheDocument()
    expect(
      screen.getByText(
        /Inscreva-se para ser o primeiro a saber sobre novos desafios :\)/i
      )
    ).toBeInTheDocument()

    expect(screen.getByRole('textbox', { name: /email/i })).toBeInTheDocument()
    expect(
      screen.getByRole('button', { name: /inscrever/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  describe('when NewsletterForm is submitted', () => {
    describe('when it succeeds', () => {
      beforeAll(() => {
        saveEmail.mockImplementation(
          jest.fn(() => Promise.resolve({ data: 'ok' }))
        )
      })

      it('should show a success message', async () => {
        render(<Newsletter />)

        const email = 'user@email.com'

        const emailField = screen.getByRole('textbox', { name: /email/i })
        const submitButton = screen.getByRole('button', { name: /inscrever/i })

        userEvent.type(emailField, email)
        await waitFor(() => {
          expect(emailField).toHaveValue(email)
        })

        userEvent.click(submitButton)
        await waitFor(() => {
          expect(
            screen.getByText(
              'Feito! Você será o primeiro a saber sobre novos desafios :)'
            )
          ).toBeInTheDocument()
        })
      })
    })

    describe('when it does not succeeds', () => {
      beforeAll(() => {
        saveEmail.mockImplementation(
          jest.fn(() => Promise.resolve({ error: 'ok' }))
        )
      })

      it('should show an error message', async () => {
        render(<Newsletter />)

        const email = 'user@email.com'

        const emailField = screen.getByRole('textbox', { name: /email/i })
        const submitButton = screen.getByRole('button', { name: /inscrever/i })

        userEvent.type(emailField, email)
        await waitFor(() => {
          expect(emailField).toHaveValue(email)
        })

        userEvent.click(submitButton)
        await waitFor(() => {
          expect(
            screen.getByText('Opa, algo deu errado! Pode tentar novamente? :c')
          ).toBeInTheDocument()
        })
      })
    })
  })
})
