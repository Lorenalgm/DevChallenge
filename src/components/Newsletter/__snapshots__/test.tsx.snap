// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Newsletter /> should render correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-margin-start: auto;
  margin-inline-start: auto;
  -webkit-margin-end: auto;
  margin-inline-end: auto;
}

@media screen and (min-width: 30em) {
  .emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media screen and (min-width: 48em) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-1 {
  height: 300px;
  width: 300px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: 0px;
}

@media screen and (min-width: 30em) {
  .emotion-2 {
    margin-left: 0px;
  }
}

@media screen and (min-width: 48em) {
  .emotion-2 {
    margin-left: var(--chakra-space-6);
  }
}

@media screen and (min-width: 62em) {
  .emotion-2 {
    margin-left: var(--chakra-space-10);
  }
}

.emotion-3 {
  font-family: var(--chakra-fonts-heading);
  font-weight: var(--chakra-fontWeights-bold);
  font-size: var(--chakra-fontSizes-2xl);
  line-height: 1.33;
}

@media screen and (min-width: 30em) {
  .emotion-3 {
    font-size: var(--chakra-fontSizes-3xl);
  }
}

@media screen and (min-width: 48em) {
  .emotion-3 {
    font-size: var(--chakra-fontSizes-3xl);
    line-height: 1.2;
  }
}

@media screen and (min-width: 62em) {
  .emotion-3 {
    font-size: var(--chakra-fontSizes-4xl);
  }
}

.emotion-4 {
  color: var(--chakra-colors-primary-500);
  font-size: var(--chakra-fontSizes-md);
}

@media screen and (min-width: 30em) {
  .emotion-4 {
    font-size: var(--chakra-fontSizes-lg);
  }
}

@media screen and (min-width: 48em) {
  .emotion-4 {
    font-size: var(--chakra-fontSizes-lg);
  }
}

@media screen and (min-width: 62em) {
  .emotion-4 {
    font-size: var(--chakra-fontSizes-xl);
  }
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: var(--chakra-space-6);
}

.emotion-5>*:not(style)~*:not(style) {
  margin-top: var(--chakra-space-4);
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-6 {
  width: 100%;
  position: relative;
}

.emotion-7 {
  width: 100%;
  min-width: 0px;
  outline: 2px solid transparent;
  outline-offset: 2px;
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  font-size: var(--chakra-fontSizes-lg);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  height: var(--chakra-sizes-12);
  border-radius: var(--chakra-radii-md);
  border: 2px solid;
  border-color: var(--chakra-colors-transparent);
  background: var(--chakra-colors-white);
  color: var(--chakra-colors-black);
}

.emotion-7:hover,
.emotion-7[data-hover] {
  background: var(--chakra-colors-whiteAlpha-900);
}

.emotion-7[aria-readonly=true],
.emotion-7[readonly],
.emotion-7[data-readonly] {
  box-shadow: none!important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-7[disabled],
.emotion-7[aria-disabled=true],
.emotion-7[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-7[aria-invalid=true],
.emotion-7[data-invalid] {
  border-color: #FC8181;
}

.emotion-7:focus,
.emotion-7[data-focus] {
  background: var(--chakra-colors-whiteAlpha-900);
  border-color: #fbbf24;
  box-shadow: 0 0 0 3px #f1d537;
}

.emotion-7::-webkit-input-placeholder {
  color: var(--chakra-colors-blackAlpha-500);
}

.emotion-7::-moz-placeholder {
  color: var(--chakra-colors-blackAlpha-500);
}

.emotion-7:-ms-input-placeholder {
  color: var(--chakra-colors-blackAlpha-500);
}

.emotion-7::placeholder {
  color: var(--chakra-colors-blackAlpha-500);
}

.emotion-8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  width: 100%;
  line-height: 1.2;
  border-radius: var(--chakra-radii-md);
  font-weight: var(--chakra-fontWeights-bold);
  height: var(--chakra-sizes-12);
  min-width: var(--chakra-sizes-12);
  font-size: var(--chakra-fontSizes-lg);
  -webkit-padding-start: var(--chakra-space-6);
  padding-inline-start: var(--chakra-space-6);
  -webkit-padding-end: var(--chakra-space-6);
  padding-inline-end: var(--chakra-space-6);
  background: #5c0d78;
  color: var(--chakra-colors-white);
}

.emotion-8:focus,
.emotion-8[data-focus] {
  box-shadow: 0 0 0 3px #a523d4;
}

.emotion-8[disabled],
.emotion-8[aria-disabled=true],
.emotion-8[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-8:hover,
.emotion-8[data-hover] {
  background: #721494;
}

.emotion-8:hover[disabled],
.emotion-8[data-hover][disabled],
.emotion-8:hover[aria-disabled=true],
.emotion-8[data-hover][aria-disabled=true],
.emotion-8:hover[data-disabled],
.emotion-8[data-hover][data-disabled] {
  background: var(--chakra-colors-whiteAlpha-200);
}

.emotion-8:active,
.emotion-8[data-active] {
  background: #D427D8;
}

<div
  class="emotion-0"
>
  <div
    class="lf-player-container"
  >
    <div
      class="emotion-1"
      id="lottie"
      style="background: transparent; margin: 0px auto; outline: none; overflow: hidden;"
    />
  </div>
  <div
    class="emotion-2"
  >
    <h3
      class="chakra-heading emotion-3"
    >
      Seja notificado sobre novos desafios!
    </h3>
    <p
      class="chakra-text emotion-4"
    >
      Inscreva-se para ser o primeiro a saber sobre novos desafios :)
    </p>
    <form
      class="chakra-stack emotion-5"
    >
      <div
        class="chakra-form-control emotion-6"
        role="group"
      >
        <input
          aria-label="Email"
          aria-required="true"
          class="chakra-input emotion-7"
          id="email"
          placeholder="Email"
          required=""
          type="email"
          value=""
        />
      </div>
      <button
        class="chakra-button emotion-8"
        type="submit"
      >
        Inscrever
      </button>
    </form>
  </div>
</div>
`;
